[DATA-790] Support for ingesting URDFs to generate model configurations

spieswl opened the pull request:
Deprecated comment

raybjork commented on referenceframe/urdf.go:
```diff
@@ -0,0 +1,258 @@
+		geoRPY := convStringAttrToFloats(jointElem.Origin.RPY)
+		jointPose := spatial.NewPoseFromOrientation(r3.Vector{X: geoXYZ[0], Y: geoXYZ[1], Z: geoXYZ[2]},
+			&spatial.EulerAngles{Roll: geoRPY[0], Pitch: geoRPY[1], Yaw: geoRPY[2]})
+		jointOrigins[jointName] = jointPose
```
I dont think you should need all this.  I might not be understanding something though.  Will continue to read

raybjork commented on referenceframe/urdf.go:
```diff
@@ -0,0 +1,258 @@
+		return nil, errors.Errorf("Invalid terminal link count: %d", len(terminalLinks))
+	}
+
+	// Create joint and link ordering, starting with the end effector link and going backwards
```
looks like this is mostlty the same as the code in `model_json` might be possible to make a helper function

raybjork reviewed (commented):
Looking like a good start.  I don't think I understand how the pose of the URDF entities is being used though.  Maybe we can discuss offline

spieswl commented on referenceframe/urdf.go:
```diff
@@ -0,0 +1,258 @@
+			Name    string   `xml:"name,attr"`
+			Origin  struct{
+				XMLName xml.Name `xml:"origin"`
+				RPY     string   `xml:"rpy,attr"`  // "r p y" format
```
I had tried as a []float64 and it fails to parse the XML. I suspect you need a custom decoder from my readings of the `encoding/xml` package, but I haven't investigated further yet.

raybjork replied:
cool just was curious.  your way works nice too

spieswl replied:
Otherwise, I'm with you. I'd like to have those just come in as the data structure we care about. Maybe something we could improve in the future when we have some time to experiment with the `xml` package.

biotinker commented on referenceframe/urdf.go:
```diff
@@ -0,0 +1,258 @@
+			Name    string   `xml:"name,attr"`
+			Origin  struct{
+				XMLName xml.Name `xml:"origin"`
+				RPY     string   `xml:"rpy,attr"`  // "r p y" format
```
Will poses always be RPY euler angles?

Also, are these degrees or radians? A comment on this struct should say.

spieswl replied:
Always RPY, always Euler angles, always radians, always meters. I can add all that detail to struct comments.

biotinker replied:
Yes please :)

spieswl commented:
Updated to real PR.

URDF parsing now creates a ModelConfig under the hood and populates that with URDF data (modified as necessary). This is so we can use the same `ParseConfig` method in `model_json.go`, and we can theoretically Marshal to JSON from a config generated by a URDF file. Added tests are very similar to those for ModelJSON tests: This is by design. I also added the logic to WrapperArm to detect model files that end in `.urdf` and handle them with the appropriate method from the referenceframe package.

I still need to squash the commits (I'm going to manually do that to preserve some history involving the logic that implemented a straight URDF -> Model conversion from before this review candidate. Other URDFs of interest which we could parse for testing purposes will be part of future PRs.

biotinker commented on referenceframe/model_json.go:
```diff
@@ -170,27 +176,11 @@ func (config *ModelConfig) ParseConfig(modelName string) (Model, error) {
-	// After the above loop, the transforms are in reverse order, so we reverse the list.
-	for i, j := 0, len(orderedTransforms)-1; i < j; i, j = i+1, j-1 {
-		orderedTransforms[i], orderedTransforms[j] = orderedTransforms[j], orderedTransforms[i]
+	orderedTransforms, err := sortTransforms(transforms, parentMap, eename, World)
```
Can assign directly to `model.OrdTransforms` here

biotinker commented on referenceframe/urdf.go:
```diff
@@ -0,0 +1,299 @@
+
+		// Relationship tracking
+		parentMap[jointElem.Name] = jointElem.Parent.Link
+		parentMap[jointElem.Child.Link] = jointElem.Name
```
Won't this double-assign each frame? Once when we see a frame, and once when we see its parent?

If so, we should do a check whether each of these are in the map already, and if so, we should verify that their intended value matches what it already is.

spieswl replied:
Since these relationships are assigned when striding the URDFs joints (and, by convention, there must be a link in between each joint), you end up assigning the adjacent links exactly twice*: Once as a child and once as a parent.

parent[joint1] -> link1 (first assignment)
parent[link2] -> link2 (second assignment)

*Except for the most proximal and most distal links, and there are attempts to catch the unique cases which may emerge as part of ignoring **world** links, etc.

biotinker replied:
I'm not sure I'm following; is your example incorrect with link2?

spieswl replied:
Yes, sorry that has been corrected in the comment.

biotinker replied:
Ah; so each item is assigned to be a key or value only once, because we're iterating through the joints here only.

Can you confirm we will never need to add links in the same way? There can never be a link whose parent is another link?

biotinker commented on referenceframe/urdf.go:
```diff
@@ -0,0 +1,299 @@
+			// Slightly different limits handling for continuous, revolute, and prismatic joints
+			if jointElem.Type == "continuous" {
+				thisJoint.Type = "revolute" // Currently, we treate a continuous joint as a special case of the revolute joint
+				thisJoint.Min, thisJoint.Max = utils.RadToDeg(-math.Pi), utils.RadToDeg(math.Pi)
```
Is +/- Pi really the default we want for continuous joints? Shouldn't these be +/- math.Inf?

spieswl replied:
A previous comment DID mention this was PR comment bait. :wink: I will make the change to math.Infs

biotinker commented on referenceframe/urdf.go:
```diff
@@ -0,0 +1,299 @@
+				thisJoint.Min, thisJoint.Max = utils.RadToDeg(-math.Pi), utils.RadToDeg(math.Pi)
+			} else if jointElem.Type == "prismatic" {
+				thisJoint.Min, thisJoint.Max = jointElem.Limit.Lower*1000, jointElem.Limit.Upper*1000 // from meters to mm
+			} else {
```
Let's do an explicit check for "revolute", and return an error if it's none of these three.

I know the switch statement above covers that, but it could change.

biotinker commented on referenceframe/urdf.go:
```diff
@@ -0,0 +1,299 @@
+				thisJoint.Type = "revolute" // Currently, we treate a continuous joint as a special case of the revolute joint
+				thisJoint.Min, thisJoint.Max = utils.RadToDeg(-math.Pi), utils.RadToDeg(math.Pi)
+			} else if jointElem.Type == "prismatic" {
+				thisJoint.Min, thisJoint.Max = jointElem.Limit.Lower*1000, jointElem.Limit.Upper*1000 // from meters to mm
```
We've got the *1000 magic number all over the place; can we make a `metersToMM` utility function instead?

biotinker commented on referenceframe/urdf.go:
```diff
@@ -0,0 +1,299 @@
+			childOrient, err := spatial.NewOrientationConfig(childEA.EulerAngles())
+
+			// Note the conversion from meters to mm
+			childLink.Translation = spatial.TranslationConfig{childXYZ[0] * 1000, childXYZ[1] * 1000, childXYZ[2] * 1000}
```
`metersToMM`

biotinker commented on referenceframe/urdf.go:
```diff
@@ -0,0 +1,299 @@
+			linkOrient, err := spatial.NewOrientationConfig(linkEA.AxisAngles())
+
+			// Note the conversion from meters to mm
+			thisLink.Translation = spatial.TranslationConfig{linkXYZ[0] * 1000, linkXYZ[1] * 1000, linkXYZ[2] * 1000}
```
`metersToMM`

biotinker commented on referenceframe/urdf.go:
```diff
@@ -0,0 +1,299 @@
+	// Handle links
+	for _, linkElem := range urdf.Links {
+		// Skip any world links
+		if linkElem.Name == "world" {
```
q: in URDF, is it guaranteed that something named "world" has size zero?

If not, we should check, and if it's non-zero, or has an orientation, we should error.

spieswl replied:
That's a tough question. I've never personally seen a modified world link. Note that links also do encode their own origin information (just offset data for visual, collision, or inertial elements, which, again, I have never seen).

biotinker replied:
Follow up question, won't this allow multiple things to be named "world"?

biotinker commented on referenceframe/urdf.go:
```diff
@@ -0,0 +1,299 @@
+		}
+
+		// Find matching links which already exist, take care of geometry if collision elements are detected
+		hasCollision := false
```
This can just be `hasCollision := len(linkElem.Collision) > 0`

biotinker commented on referenceframe/urdf.go:
```diff
@@ -0,0 +1,299 @@
+			hasCollision = true
+		}
+
+		for idx, prefabLink := range mc.Links {
```
What are we looking for here? `fixed` joints? Is there guaranteed to only be one?

spieswl replied:
If we've already populated a `jsonLink` while we were parsing joints. There should only be one link that has a particular name (hence the hunt for the prefabLink's name), so we should fill out data we haven't yet encountered.

biotinker commented on referenceframe/urdf.go:
```diff
@@ -0,0 +1,299 @@
+			thisLink := JsonLink{ID: linkElem.Name, Parent: World}
+
+			linkEA := spatial.EulerAngles{Roll: 0.0, Pitch: 0.0, Yaw: 0.0}
+			linkOrient, err := spatial.NewOrientationConfig(linkEA.AxisAngles())
```
This can just be `spatial.OrientationConfig{}`, no need for EA.

Add a comment saying the orientation is guaranteed to be zero

biotinker commented on referenceframe/urdf.go:
```diff
@@ -0,0 +1,299 @@
+		boxDims := convStringAttrToFloats(boxGeometry.Size)
+		geoCfg = spatial.GeometryConfig{
+			Type:              "box",
+			X:                 boxDims[0] * 1000, // from meters to mm
```
`metersToMM`

spieswl replied:
Not sure what caused a few of those comments to be hidden, but yeah I'll implement that convenience method

biotinker commented on referenceframe/urdf.go:
```diff
@@ -0,0 +1,299 @@
+			Label:             "sphere",
+		}
+	} else {
+		err = errors.Errorf("Unsupported collision geometry type detected for [ %v ] link", link.Collision[0].Name)
```
Can return this directly

tessavitabile commented on motionplan/kinematic_test.go:
```diff
@@ -389,3 +389,25 @@ IK:
+	test.That(t, err, test.ShouldBeNil)
+
+	seed := rand.New(rand.NewSource(50))
+	for i := 0; i < numTests; i++ {
```
Is the intention to change the seed for each iteration?

spieswl replied:
Not currently. We do get a different random joint configuration for each iteration of the loop, however, but for each run of the test we do get the same joint sequences.

spieswl replied:
In the `GenerateRandomConfiguration` function over in `model.go` (L55), we end up pulling a DoF-length amount of random numbers from the `seed` each time we go through the loop.

biotinker replied:
More generally:

Random number seeds in Go are deterministic, reusable sources of randomness. So if you call:
```
	seed := rand.New(rand.NewSource(50))
	fmt.Println(seed.Int())
	fmt.Println(seed.Int())
	fmt.Println(seed.Int())
```
You'll get three different numbers, but they'll be the same three numbers each time you run the code.

tessavitabile replied:
thanks!

tessavitabile commented on referenceframe/model.go:
```diff
@@ -240,3 +240,32 @@ func floatsToString(inputs []Input) string {
 }
+
+// Create an ordered list of transforms given a parent mapping, keeping an eye out for a sentinel string (World).
+func sortTransforms(unsorted map[string]Frame, parentMap map[string]string, start, finish string) ([]Frame, error) {
```
To make sure I understand, the goal is to produce an ordered list of transforms starting from `finish` (exclusive) to `unordered[start]` (inclusive), using the parent mapping. This list may not include all of the transforms in `unsorted`. Is that correct?

spieswl replied:
More or less, that is correct. I would defer to @biotinker on the detail of whether or not the intention is to hit all `unsorted` transforms. I don't believe we support branching like that in `ModelJSON`.

biotinker replied:
Agreed. That shouldn't be possible with our current usage, but since this is now a function that can be called from elsewhere, we should start making guarantees.

spieswl replied:
DATA-877 captures the concerns in this comment chain.

tessavitabile commented on referenceframe/model_json.go:
```diff
@@ -13,34 +13,40 @@ import (
+	Min    float64            `json:"min"` // in mm or degs
+}
+
+type JsonDHParam struct {
```
What is a DH param?

spieswl replied:
Denavit–Hartenberg parameters, or another way of encoding the kinematics of a given robot.

tessavitabile commented on referenceframe/urdf.go:
```diff
@@ -0,0 +1,299 @@
+		childLink := JsonLink{ID: jointElem.Child.Link, Parent: jointElem.Name}
+
+		switch jointElem.Type {
+		case "continuous", "revolute", "prismatic":
```
Can you make sure all joint types are covered by unit tests? It looks like "prismatic" is not.

spieswl replied:
I added an additional URDF file that represents an X-Y gantry composed of only prismatic joints and added parsing of that to the current URDF tests.

tessavitabile commented on referenceframe/urdf.go:
```diff
@@ -0,0 +1,299 @@
+		}
+	}
+
+	return mc.ParseConfig(modelName)
```
Would it make sense to return a `ModelConfig` here and wrap in a function that returns a `Model`? That would allow us to write unit tests that check the expected `ModelConfig` directly.

spieswl replied:
Yes! This signature matches a roughly equivalent method over in `model_json.go` as well, so I would be keen to update both files to follow that pattern.

spieswl replied:
I changed this behavior for the URDF code; I broke a test when I attempted to do this in `model_json`, so I paused that and just applied this change to URDF. I would still like to come back and make this behavior common.

spieswl replied:
DATA-877 captures the effort to come back and make this behavior common.

tessavitabile reviewed (commented):
Thanks, this was very helpful to read!

biotinker reviewed (approved):
LGTM

raybjork commented on referenceframe/urdf.go:
```diff
@@ -0,0 +1,313 @@
+			childXYZ := convStringAttrToFloats(jointElem.Origin.XYZ)
+			childRPY := convStringAttrToFloats(jointElem.Origin.RPY)
+			childEA := spatial.EulerAngles{Roll: childRPY[0], Pitch: childRPY[1], Yaw: childRPY[2]}
+			childOrient, err := spatial.NewOrientationConfig(childEA.EulerAngles())
```
[nit ] You don't need to call `EulerAngles` here

raybjork commented on referenceframe/urdf.go:
```diff
@@ -0,0 +1,313 @@
+			linkXYZ := convStringAttrToFloats(jointElem.Origin.XYZ)
+			linkRPY := convStringAttrToFloats(jointElem.Origin.RPY)
+			linkEA := spatial.EulerAngles{Roll: linkRPY[0], Pitch: linkRPY[1], Yaw: linkRPY[2]}
+			linkOrient, err := spatial.NewOrientationConfig(linkEA.AxisAngles())
```
same here.  Actually it seems like there is a lot of duplicate code here and the above case that could be abstracted away

raybjork commented on referenceframe/model_json.go:
```diff
@@ -13,34 +13,40 @@ import (
+	RawFrames    []FrameMapConfig `json:"frames"`
+}
+
+type JsonLink struct {
```
Depending on who wins the PR race this is cause conflicts for either you or Peter. He is renaming these to LinkConfig and JointConfig I think.  Might help cut down on merge conflicts for one of you

raybjork commented on referenceframe/errors.go:
```diff
@@ -24,3 +24,13 @@ func NewFrameAlreadyExistsError(frameName string) error {
+
+// NewCircularReferenceError returns an error indicating that a circular path exists somewhere between the end effector and the world.
+func NewCircularReferenceError() error {
+	return errors.Errorf("infinite loop finding path from end effector to world")
```
Error functions without arguments can just encode the error as a variable as in line 6 of this file

raybjork commented on referenceframe/urdf.go:
```diff
@@ -0,0 +1,313 @@
+			}
+		}
+
+		// In the event the link does not already exist in the ModelConfig, we will have to generate it now
```
when would this already exist in the ModelConfig?

spieswl replied:
On L131, we create a child link prior to parsing through the URDF Links themselves.

raybjork commented on referenceframe/urdf.go:
```diff
@@ -0,0 +1,313 @@
+			OrientationOffset: *geomOx,
+			Label:             "box",
+		}
+	} else if sphereGeometry.Radius > 0 {
```
The way the geometry config parser is written this should be a lot simpler to do.  You should be able to just shove everything in the same config struct and have it figure it out, since this logic is replicated there

spieswl replied:
Yeah, I'm not going to mess with that at the moment, though.
